packages/php-wasm/compile/php/php_wasm.c

```diff
@@ -3,6 +3,26 @@
  *
  * This file abstracts the entire PHP API with the minimal set
  * of functions required to run PHP code from JavaScript.
+ *
+ * !! NOTE FOR FUTURE MAINTAINERS !!
+ *
+ * When debugging memory errors, it's important to understand the behavior of the stack in the Emscripten environment:
+ *
+ * - The stack does not grow automatically. If the stack runs out of space, it will not be expanded automatically, unlike in traditional C programs built with compilers like GCC or Clang.
+ * - The stack size is limited. The default stack size in Emscripten is considerably smaller than in a typical C environment. There have been discussions about reducing it even further.
+ *
+ * When passing in long strings from JS, Use the heap instead of the stack to hold large chunks of data. Here's how:
+ *
+ * // Set the memory
+ * const size = Module.lengthBytesUTF8(body);
+ * const addr = Module._malloc(size + 1);
+ * Module.stringToUTF8(body, addr, size + 1);
```

Comment:
I've seen invalid Unicode strings too often to expect this not to cause _some_ kind of problem. JS makes it particularly easy to split surrogate pairs.

can we document what happens when this is provided with invalid Unicode?

for an example of an invalid string we can pass the JS value `'\ud83c'`, which is one half of a UTF-16 surrogate pair.

Reply:
@dmsnell Does this handle the case you're worried about?

https://github.com/emscripten-core/emscripten/blob/85c099ab519d574c3ce1f6d3603eb74aabbdf604/src/library_strings.js#L235-L254

Reply:
thanks for the link, @seanmorris - I'm not sure what case this is supposed to address, I don't think that having code in another repository has much documentation value here. I'm only examining this from the perspective of what we're writing, which is a guide for how to transfer strings from JavaScript to WASM, but we don't call out one important edge case which could lead to memory problems or corruption.

what happens if someone sends an invalid Unicode string from JavaScript? will `lengthBytesUTF8` return the appropriate number of bytes needed for `stringToUTF8`? what happens in the case of invalid Unicode?

here's an idea:

```js
 /*
 * // Set the memory
 * const size = Module.lengthBytesUTF8(body);
 * const addr = Module._malloc(size + 1);
 * Module.stringToUTF8(body, addr, size + 1);
 *
 * Note: If provided an invalid string this conversion will be lossy.
 *       Invalid code points will be converted to U+FFFD and
 *       `Module.lengthBytesUTF8` will report the number of bytes after
 *       converting those code points. E.g. `a\ud83cb` turns into
 *       the string `a�b` and the length is 1 + 3 + 1 = 5.
```

now I haven't checked if this is how `emscripten`'s library is doing the conversion: that's the part I was hoping you could confirm and document in place so that everyone coming to this code wouldn't have to.

Reply:
even better, it might be instructive to list as an example a common case that creates these invalid strings.

```js
/*
 * …
 * E.g. Consider a string split inside a surrogate pair boundary.
 * 
 *     `'I feel 😊.'.slice(0, 8)`
 * 
 * We might expect this to occupy 8 bytes because we split
 * the string at 8 characters, or to occupy 11 bytes because
 * we expected to get the emoji as the 8th character, and it
 * requires four bytes in UTF8, but instead we invalidated
 * the string and receive `I feel �`, which takes a total of
 * 10 bytes in UTF8: 7 to encode `I feel ` and then 3 to
 * encode the replacement character U+FFFD.
 *      length 7 + 3 = 10 instead of length 8 or length 11
```

Reply:
@dmsnell ADoes that regard how Javascript will handle the string, prior to being passed to `lengthBytesUTF8` & `stringToUTF8`, or is there something that those functions might do to a string that you're worried might cause issues?

Reply:
@dmsnell Actually it looks like in that case, the emoji will be split into `0xd83d` & `0xde0a`, each of which occupy 4 bytes in this context: https://jsfiddle.net/fog8ewr4/

How do you expect this to trip someone up, though? Is it specific to this context or is this just JS's weird string handling in general?

Reply:
@dmsnell How does this look?

https://github.com/WordPress/wordpress-playground/blob/5db03ef4c6cc7001acee97717f62cb160bc7710d/packages/php-wasm/compile/php/php_wasm.c#L28-L36

Reply:
Thanks @seanmorris for confirming the behaviors; the results are as I was worried about and are important for why we want to document this. As you linked to in [the httpget issue](https://github.com/jsoftware/j-playground/issues/80#issuecomment-1140054822) there's a chance for memory corruption or overflow because of an over-simplification of string handling.

> Does that regard how Javascript will handle the string, prior to being passed to lengthBytesUTF8 & stringToUTF8

So what's happening here is that JavaScript is fine operating on invalid Unicode strings. Internally, JavaScript strings are stored as something that's _like_ UTF-16, but actually it isn't totally because it allows creating strings that cannot be represented in Unicode. They allow an arbitrary stream of bytes to be a string, which mostly works in UTF-16 and _does_ work in UCS-2, but in order to support extended character ranges in Unicode, UTF-16 said that the code points referenced by each surrogate half are invalid and cannot be represented. So U+D83C U+DC00 is an invalid sequence of code points, but the UTF-16 sequence of bytes 0xd83cdc00 _is_ valid and converts to U+1F000. This is why they cannot be split, because you cannot represent U+D83C in UTF-8 or even abstractly in Unicode.

All of this is to say that JavaScript is fine working with these things; the problem is the conversion to UTF-8 bytes.

https://jsfiddle.net/fog8ewr4/

> How do you expect this to trip someone up, though? Is it specific to this context or is this just JS's weird string handling in general?

Can you figure it out by reviewing the confirmation work you posted in the jsfiddle?

We've confirmed that `lengthBytesUTF8` and `stringToUTF8` appear to operate on the same conversion algorithm, so that the memory allocation should be safe, but we've also confirmed that this will corrupt strings such that we can't pass them safely back and forth across the divide.

By the way, the 11 from `lengthBytesUTF8` doesn't seem to be coming from "counted as 4 bytes in this context," but if you look more closely, you will see something else is going on. For one, if it _were_ 4 bytes for those characters _each_ we'd expect not 11 bytes, but 15 bytes.

Instead, the split surrogate here is being converted to U+FFFD, which takes 3 bytes in UTF8 and then `lengthBytesUTF8` adds one extra space for a terminating null byte.

So our comment needs to communicate that if people split surrogates then the string will be corrupted in a lossy way. This doesn't only happen at the end, by the way, because we can find strings in JavaScript with split surrogates in the middle; often in practice the result of more naive string handling. `>>\ud83c<<`